spec/lib/msf/core/encoded_payload_spec.rb

```diff
@@ -17,6 +17,11 @@
       module_type: 'encoder',
       modules_path: modules_path,
     )
+    allow(framework.encoders).to receive(:each_module_ranked)
+                                   .and_yield('x86/shikata_ga_nai', framework.encoders['x86/shikata_ga_nai'])
```

Comment:
Is there any additional context to this change

Just wondering if it's a a workaround for an implementation change that we're fixing in tests? 👀

Reply:
sure, so the test was originally just loading these 4 modules into the `encoders` module set, and since those were the only 4 loaded only they would be available, I'm just updating the mocking to only have these same 4 modules being returned from `each_module_ranked` which now uses the metadata to get its list of modules rather than from what modules are loaded into the module set

Reply:
Oh so `expect_to_load_module_ancestors` actually has the side effect of forcing modules to load

I don't feel super comfortable having the `each_module_ranked` call mocked here, as it feels like the existing test is verifying a lot of metasploit (if we're treating this as more of an integration test). Not a hard blocker for me, it's just we don't have existing tests for `each_module_ranked` from what I can see from a 3 second look 🕵️

Reply:
would you be more okay with mocking out the call to getting the metadata instead?

Reply:
What are your thoughts on this?

```diff
diff --git a/lib/msf/core/module_set.rb b/lib/msf/core/module_set.rb
index 175a32adc4..7ad1207411 100644
--- a/lib/msf/core/module_set.rb
+++ b/lib/msf/core/module_set.rb
@@ -274,7 +274,7 @@ class Msf::ModuleSet < Hash
   # @return [Array<Array<String, Class>>] Array of arrays where the inner array is a pair of the module reference name
   #   and the module class.
   def rank_modules
-    module_metadata.sort_by { |refname, _metadata| module_rank(refname) }.reverse!
+    module_metadata.sort_by { |refname, _metadata| "#{module_rank(refname)}-#{refname}" }.reverse!
   end
 
   # Retrieves the rank from a loaded, not-yet-loaded, or unloadable Metasploit Module.
diff --git a/spec/lib/msf/core/encoded_payload_spec.rb b/spec/lib/msf/core/encoded_payload_spec.rb
index 88f9dcb823..cbea72498d 100644
--- a/spec/lib/msf/core/encoded_payload_spec.rb
+++ b/spec/lib/msf/core/encoded_payload_spec.rb
@@ -5,23 +5,28 @@ RSpec.describe Msf::EncodedPayload do
   include_context 'Msf::Simple::Framework#modules loading'
 
   before do
+    ancestor_reference_names = [
+      # Excellent rank
+      'x86/shikata_ga_nai',
+      # Great rank
+      'x86/call4_dword_xor',
+      'x86/xor_dynamic',
+      'generic/none',
+    ]
     expect_to_load_module_ancestors(
-      ancestor_reference_names: [
-        # Excellent rank
-        'x86/shikata_ga_nai',
-        # Great rank
-        'x86/call4_dword_xor',
-        'x86/xor_dynamic',
-        'generic/none',
-        ],
+      ancestor_reference_names: ancestor_reference_names,
       module_type: 'encoder',
       modules_path: modules_path,
     )
-    allow(framework.encoders).to receive(:each_module_ranked)
-                                   .and_yield('x86/shikata_ga_nai', framework.encoders['x86/shikata_ga_nai'])
-                                   .and_yield('x86/xor_dynamic', framework.encoders['x86/xor_dynamic'])
-                                   .and_yield('x86/call4_dword_xor', framework.encoders['x86/call4_dword_xor'])
-                                   .and_yield('generic/none', framework.encoders['generic/none'])
+
+    # Improve test performance - return only the test modules that we're interested in
+    allow(framework.encoders).to receive(:rank_modules).and_wrap_original do |original, *args|
+      ranked_modules = original.call(*args)
+
+      ranked_modules.select do |ref_name, _metadata|
+        ancestor_reference_names.include?(ref_name)
+      end
+    end
   end
 
   let(:ancestor_reference_names) {

```

🤔

Reply:
totally agree it's an improvement, have applied the patch